reverse_proxy.go

```diff
@@ -534,23 +546,43 @@ func (p *ReverseProxy) WrappedServeHTTP(rw http.ResponseWriter, req *http.Reques
 
 	var res *http.Response
 	var err error
-	if breakerEnforced {
-		log.Debug("ON REQUEST: Breaker status: ", breakerConf.CB.Ready())
-		if breakerConf.CB.Ready() {
-			res, err = transport.RoundTrip(outreq)
-			if err != nil {
-				breakerConf.CB.Fail()
-			} else if res.StatusCode == 500 {
-				breakerConf.CB.Fail()
+
+	// Handle gRPC Proxy
+	log.Debug("===> GRPC Server service name: ", p.TykAPISpec.CustomMiddleware.GRPCProxy.Name)
+	if p.TykAPISpec.CustomMiddleware.GRPCProxy.Name != "" {
+		log.Debug("====> Calling: ", outreq.URL.String())
+		if breakerEnforced {
+			if breakerConf.CB.Ready() {
+				rec := httptest.NewRecorder()
```

Comment:
Are we sure that using `httptest` outside of a test is safe?

Reply:
I do not know... I haven't found a better way to dump the request so we can use it after.

Reply:
So if we're using the grpc proxy, we're buffering the entire response? That seems pretty bad to me. If someone wishes to use a grpc API that writes tons of data or writes data incrementally or over time, this won't work at all.

I assume you did this because you only have `ServeHTTP` to use from grpc. We should be calling that directly with the user's `ResponseWriter`, instead of reading the whole response into memory first like this.

Note that this is different from what `RoundTrip` does - that one does the request, but it doesn't buffer the entire response body into memory.

Reply:
Tyk has response processors - that process the response on the way back. We could completely bypass those but then the cache will also be unavailable.

What we basically need is a copy of the response object to make that work.

I don't know how else to proceed, I don't know enough about RoundTrip.

Reply:
We can leave things as is, but it's going to be a bit limited with the buffering. I guess we can address that if it ever becomes a concern.

Reply:
I can't see another way of grabbing the response, unless we can write a matching interface that will partially buffer like RoundTrip does... but that's beyond my ken.